lib/Doctrine/DBAL/Driver/IBMDB2/DB2Statement.php

```diff
@@ -60,16 +60,18 @@ class DB2Statement implements \IteratorAggregate, Statement
      *
      * @var array
      */
-    static private $_typeMap = array(
-        \PDO::PARAM_INT => DB2_LONG,
-        \PDO::PARAM_STR => DB2_CHAR,
-    );
+    static private $_typeMap;
 
     /**
      * @param resource $stmt
      */
     public function __construct($stmt)
     {
+        self::$_typeMap = [
+            \PDO::PARAM_INT => DB2_LONG,
+            \PDO::PARAM_STR => DB2_CHAR,
```

Comment:
That's not something I anticipated and it really sucks (my bad). The static type map shouldn't be updated upon each object instantiation.

What if we take another route: right now, `getIterator()` is implemented exactly the same in all statements and we test all implementations with one test. What if instead, we move it out of the statements into a separate class and test it with a statement mock. In this case we'll be able to mock the statement interface, not test all implementations.

```php
class StatementIterator implements IteratorAggregate
{
    public function getIterator()
    {
        while (false !== ($result = $this->stmt->fetch())) {
            yield $result;
        }
    }
}

class WhateverStatement
{
    public function getIterator()
    {
        return new StatementIterator($this);
    }
}
```

The test logic will remain the same with the only exception that we'll be testing the statement iterator, not the statement classes.

Reply:
Ok I've made this change. The testIterationCallsFetchOncePerStep test doesn't pass. Without calling next (which I agree isn't right) it doesn't ever get into the callback of the statement mock. Any ideas?